File: leetcode-everyday/排序算法/347-前 K 个高频元素/347-前 K 个高频元素.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
 var topKFrequent = function(nums, k) {
  const counts = new Map();
  let max_count = 0;
  for(let num of nums) {
      if(counts.has(num)) {
          counts.set(num, counts.get(num) + 1);
      } else {
          counts.set(num, 1);
      }
      max_count = Math.max(max_count, counts.get(num));
  }

  const buckets = new Array(max_count+1).fill().map(() => []);    
  for(let [num, count] of counts) {
      buckets[count].push(num);
  }

  const ans = [];
  for(let i = max_count; i >= 0 && ans.length !== k; i--) {
      for(let num of buckets[i]) {
          ans.push(num);
          if(ans.length === k) {
              break;
          }
      }
  }
  return ans;
};

File: leetcode-everyday/排序算法/347-前 K 个高频元素/347-前 K 个高频元素.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        // [1,1,1,2,2,3]
        unordered_map<int, int> counts;
        int max_count = 0;
        for(const int & num : nums) {
            max_count = max(max_count, ++ counts[num]);
        }
        // {
        //     1: 3,
        //     2: 2,
        //     3: 1
        // }
        // 3

        vector<vector<int>> buckets(max_count + 1);
        // [[], [], [], []]
        for(const auto & p : counts) {
            buckets[p.second].push_back(p.first);
        }
        // [[], [3], [2], [1]]

        vector<int> ans;
        for(int i = max_count; i >= 0 && ans.size() < k; --i) {
            for(const int & num: buckets[i]) {
                ans.push_back(num);
                if(ans.size() == k) {
                    break;
                }
            }
        }
        return ans;
    }
};